### PR TITLE
[WIP] build: added support for building using 'provided' runtime.

### PR DESCRIPTION
**Pull request by mster**
Issue #1099

*Description of changes:*

Allow for the building of a lambda function when specifying `Runtime: provided`, with a corresponding layer ARN. SAM will automagically determine the runtime by determining the workflow configuration based off of the project manifest.

For example, with the sample `sam-app`:
```yml
...
      Runtime: provided
      Layers:
        - <YOUR-PROVIDED-RUNTIME-ARN>
...
```

* Added a `ManifestWorkflowSelector` to add support for building using the `provided` runtime flag, as suggested [here](https://github.com/awslabs/aws-sam-cli/issues/1099#issuecomment-488877333). 
* Added logging to inform user that the `provided` runtime is being used. 
* Fails when attempting to build Lambda Function within a container. (can't determine an appropriate image to pull)

*Checklist:*

- [x] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [x] Write unit tests
- [-] Write/update functional tests
- [x] Write/update integration tests
- [x] `make pr` passes
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samcli/lib/build/workflow_config.py [hunk 1]
```diff
@@ -111,13 +111,23 @@ def get_workflow_config(runtime, code_dir, project_dir):
             JAVA_KOTLIN_GRADLE_CONFIG._replace(executable_search_paths=[code_dir, project_dir]),
             JAVA_MAVEN_CONFIG
         ]),
+
+         "provided": ManifestWorkflowSelector([
+            PYTHON_PIP_CONFIG,
+            RUBY_BUNDLER_CONFIG,
+            DOTNET_CLIPACKAGE_CONFIG,
+            NODEJS_NPM_CONFIG,
+        ]),
```
Comment (sanathkr) [inline samcli/lib/build/workflow_config.py]: Is there any reason why you did not add other builders?

**Comment by sanathkr** [comment]
Can you explain how you test this?

File: samcli/lib/build/workflow_config.py (hunk 1 above)
Comment (mster) [reply samcli/lib/build/workflow_config.py]: No particular reason, mainly just apprehension for testing.

File: tests/unit/lib/build_module/test_workflow_config.py [hunk 2]
```diff
@@ -70,6 +70,30 @@ def test_must_work_for_java(self, runtime, build_file, dep_manager, os_mock):
...
+        self.assertEquals(result.language, language)
+        self.assertEquals(result.dependency_manager, dep_manager)
+        self.assertEquals(result.application_framework, None)
+        self.assertEquals(result.manifest_name, build_file)
+
+        if dep_manager == "gradle":
+            self.assertEquals(result.executable_search_paths, [self.code_dir, self.project_dir])
+        else:
+            self.assertIsNone(result.executable_search_paths)
+
```
Comment (mster) [inline tests/unit/lib/build_module/test_workflow_config.py]: @sanathkr Would this be sufficient?

Comment (TheSriram) [reply tests/unit/lib/build_module/test_workflow_config.py]: fair enough, the appropriate builder gets selected. A integration test for this, would be great as well.

**Comment by mster** [comment]
> This is a great feature! thanks for adding this. The PR mentions the "appropriate layer ARN", is this left as an exercise to the user of the CLI to figure out? or should we mention that for example "Java runtime layer ARN needs to be provided since a java builder was picked on the basis of the manifest file?"

Excellent point. I do feel as though it should be up to the user to figure this out, however a log message during the build phase would be a useful addition. I will add that, let me know what you think.

**Comment by mster** [comment]
@TheSriram for example:

```sh
2019-08-12 14:51:25 Building resource using 'provided' runtime.
2019-08-12 14:51:25 nodejs workflow detected. Please make sure you are using the appropriate layer ARN.
```

**Comment by TheSriram** [comment]
@mster Awesome, thanks for updating the PR. I think the log message definitely makes sense. I still have some concerns on if we need to provide a mechanism for a `true` provided runtime that does not have a native builder yet? Do you have any thoughts on how we could integrate that? like a build plugin? (Thinking out loud here)

**Comment by mster** [comment]
> have some concerns on if we need to provide a mechanism for a `true` provided runtime that does not have a native builder yet? Do you have any thoughts on how we could integrate that? like a build plugin? (Thinking out loud here)
 
What about requiring something in the `template.yaml` that specifies the builder to be used? 

```
...
      Runtime: provided
      Layers:
        - <provided-runtime-arn>
      Builder: <valid-native-builder>
...
```

Not sure on how to accomplish without the native builder.

**Comment by jfuss** [comment]
Closing this as the issue linked has been closed and has been supported in SAM CLI. See https://github.com/aws/aws-sam-cli/issues/1099#issuecomment-661401086 for details.

**Pull request closed without merging**